### ISSUE TITLE
Request: Make `Server` singleton

### ISSUE THREAD

**Issue by Levivb** [upgate/laravel-jsonrpc#14]
Hey,

See: https://github.com/upgate/laravel-jsonrpc/blob/master/src/ServiceProvider/JsonRpcServerServiceProvider.php#L20

Right now when resolving the `Server` via Dependency Injection in Controllers, a new instance is created for each injection.
This makes it impossible to configure the `Server` in separate files since a new instance is given for each DI.

I managed to fix it by adding this package to `dont-discover` in `composer.json`:
```json
"dont-discover": [
        "upgate/laravel-jsonrpc"
      ]
```

And registering a separate Service Provider which binds the classes as singletons:
```php
class JsonRpcServerServiceProvider extends ServiceProvider
{
    public function register(): void
    {
        $this->app->singleton(RouteRegistryInterface::class, Router::class);
        $this->app->singleton(RequestFactoryInterface::class, RequestFactory::class);
        $this->app->bind(RouteDispatcherInterface::class, RouteDispatcher::class);
        $this->app->singleton(MiddlewareDispatcherInterface::class, MiddlewarePipelineDispatcher::class);
        $this->app->singleton(ServerInterface::class, Server::class);
    }
}
```

And separated code to setup routing:
```php
final class JobApplicationJsonRpcRegistrator implements Registrator
{
    /** @var JsonRpcServer */
    private $server;

    public function __construct(JsonRpcServer $server)
    {
        $this->server = $server;
    }

    public function register(): void
    {
        $this->server
            ->router()
            ->bindController('job-application', JobApplicationController::class);
    }
}
```
```php
final class VacancyJsonRpcRegistrator implements Registrator
{
    /** @var JsonRpcServer */
    private $server;

    public function __construct(JsonRpcServer $server)
    {
        $this->server = $server;
    }

    public function register(): void
    {
        $this->server
            ->router()
            ->bindController('vacancy', VacancyController::class);
    }
}
```

Would it be possible to bind the `ServerInterface` as singleton to the concrete `Server` as is shown in the `JsonRpcServerServiceProvider` snippet above?

Let me know what you think :)

**Comment by kbaryshnikov** [comment]
This is a breaking change, so I'm not really sure about it.

I see no good reason for not making it a singleton though.

Please see if #18 works for you.

**Comment by kbaryshnikov** [comment]
Please see if #19 works for you.

**Comment by Levivb** [comment]
Thanks for the fast response! 👍

This seems to work perfectly!

One thing to keep in mind is that the `RouteDispatcher` is stateful since it stores the string passed to `setControllerNamespace`. So working with a singleton could introduce unexpected behaviour when multiple namespaces are set - depending on how the developer uses this package.

In my opinion working with FQN's is preferred over storing some namespace separately from the controller name.
So `->bindController('foo', FooController::class);` looks way nicer then
vs `->setControllerNamespace('Some\Namespace')->bindController('foo', 'FooController')`
Added benefit: automatic refactoring in your IDE takes care of updating the former, while the later won't be refactored along. But that's just my two cents :)

Looking over the old code.... I don't think running `setControllerNamespace` multiple times with different values worked before as well since you'd always have only one working `JsonRpcServer` and thus only one `RouteDispatcher.
So no functionality is lost on that part 😀

**Comment by Levivb** [comment]
Ow, and `src/ServiceProvider/JsonRpcServerFactory.php` makes more sense located in `src/Server/` since it isn't a Service Provider but a factory :)

**Comment by kbaryshnikov** [comment]
> RouteDispatcher is stateful since it stores the string passed to setControllerNamespace

Yeah. That's why `ServerFactory::make()` creates a new instance of `RouteDispatcher` on every call.

> In my opinion working with FQN's is preferred over storing some namespace separately from the controller name.

You can always do that. `setControllerNamespace()` is optional.

I've updated the test suite to reflect that.

> ...ServerFactory makes more sense located in src/Server

Well, yeah, I could not decide, now I can. Moved.

Please see updated #19

**Comment by Levivb** [comment]
> Yeah. That's why ServerFactory::make() creates a new instance of RouteDispatcher on every call.

Ah, I see. Nice!

> You can always do that. setControllerNamespace() is optional.

Yea, I know :) It's good to provide `setControllerNamespace` for those that prefer otherwise


> Please see updated #19

Cool! Looks good :)

**Comment by kbaryshnikov** [comment]
Merged #19 to master.

**Comment by kbaryshnikov** [comment]
https://github.com/upgate/laravel-jsonrpc/releases/tag/0.5.0
